gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+        }
+    }
+  while (ok);
+
+  return NULL;
+}
+
+static gpointer do_decompress (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean ok = TRUE;
+
+  do
+    {
+      ok = eos_reformatter_decompress (reformatter, reformatter->decomp_queue, reformatter->write_queue);
+    }
+  while (ok);
+
+  return NULL;
+}
+
+static gpointer do_read_and_decompress (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean read_ok = TRUE;
+  gboolean decomp_ok = TRUE;
+
+  do
+    {
+      if (read_ok)
+        read_ok = eos_reformatter_read (reformatter, reformatter->decomp_queue);
+
+      decomp_ok = eos_reformatter_decompress (reformatter, reformatter->decomp_queue, reformatter->write_queue);
+    } while (read_ok || decomp_ok);
+
+  return NULL;
+}
+
+static gpointer do_write (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER (data);
+  gboolean ok = TRUE;
+
+  do
+    {
+      ok = eos_reformatter_write (reformatter, reformatter->write_queue);
+    } while (ok);
+
+  return NULL;
+}
+
+
+/* Public API */
+
+gdouble
+eos_reformatter_get_progress (EosReformatter *reformatter)
+{
+  gdouble progress = 0.0;
+  g_object_get (reformatter, "progress", &progress, NULL);
+  return progress;
+}
+
+GError *
+eos_reformatter_get_error (EosReformatter *reformatter)
+{
+  return reformatter->error;
+}
+
+guint64
+eos_reformatter_get_usec_remaining (EosReformatter *reformatter)
+{
+  if (reformatter->estimator == NULL)
+    return 0;
+
+  return gdu_estimator_get_usec_remaining (reformatter->estimator);
+}
+
+guint64
+eos_reformatter_get_bytes_per_sec (EosReformatter *reformatter)
+{
+  if (reformatter->estimator == NULL)
+    return 0;
+
+  return gdu_estimator_get_bytes_per_sec (reformatter->estimator);
+}
+
+gboolean
+eos_reformatter_reformat (EosReformatter *reformatter)
+{
+  g_return_val_if_fail (reformatter != NULL, FALSE);
+  g_return_val_if_fail (reformatter->image != NULL, FALSE);
+  g_return_val_if_fail (reformatter->device != NULL, FALSE);
+
+  g_object_set (reformatter, "progress", 0.0, NULL);
+
+  reformatter->read_fd = eos_reformatter_prepare_read (reformatter);
+  if (reformatter->read_fd <= 0)
+    {
+      return FALSE;
+    }
+
+  if (reformatter->signature != NULL && !eos_reformatter_prepare_gpg_verify (reformatter))
+    {
+      return FALSE;
+    }
+
+  reformatter->write_fd = eos_reformatter_prepare_write (reformatter);
+  if (reformatter->write_fd <= 0)
+    {
+      return FALSE;
+    }
+
+  reformatter->estimator = gdu_estimator_new (reformatter->write_total_bytes);
+  g_signal_connect (reformatter->estimator, "notify::usec-remaining",
+                    G_CALLBACK (eos_reformatter_update_progress),
+                    reformatter);
+
+  g_warning ("%s: using %d threads", __FUNCTION__, EOS_THREADS);
+
+#if EOS_THREADS == 1
+  reformatter->write_thread = g_thread_new ("read-write", do_reformat, reformatter);
```

Comment:
Right now we (try to) make sure the threads will exit before dispose finishes, so no explicit ref is needed. If we add refs in the threads, a simple g_object_unref() would not trigger shutdown I believe.